legend-engine-xts-analytics/legend-engine-xts-analytics-quality/legend-engine-xt-analytics-quality-api/src/main/java/org/finos/legend/engine/api/analytics/DataspaceQualityAnalytics.java

```diff
@@ -0,0 +1,242 @@
+import org.finos.legend.pure.m3.coreinstance.meta.pure.metamodel.function.ConcreteFunctionDefinition;
+import org.finos.legend.pure.m3.coreinstance.meta.pure.metamodel.function.property.Property;
+import org.finos.legend.pure.m3.coreinstance.meta.pure.metamodel.type.Class;
+import org.finos.legend.pure.m3.coreinstance.meta.pure.metamodel.type.Enum;
+import org.finos.legend.pure.m3.coreinstance.meta.pure.metamodel.type.Enumeration;
+import org.finos.legend.pure.runtime.java.compiled.execution.CompiledExecutionSupport;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Api(tags = "Analytics - Quality")
+@Path("pure/v1/analytics/quality")
+public class DataspaceQualityAnalytics
+{
+    private final ModelManager modelManager;
+    private List<String> messages;
+
+    public DataspaceQualityAnalytics(ModelManager modelManager)
+    {
+        this.modelManager = modelManager;
+    }
+
+    @POST
+    @Path("check")
+    @ApiOperation("Checks the quality of provided Data space")
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response checkDataSpaceConstraints(PureModelContextData pureModelContextData)
+    {
+        messages = new ArrayList<>();
+        PureModel pureModel = this.modelManager.loadModel(pureModelContextData, null, null, null);
+
+        List<PackageableElement> classes = pureModelContextData.getElements().stream().filter(e -> e instanceof org.finos.legend.engine.protocol.pure.v1.model.packageableElement.domain.Class).collect(Collectors.toList());
+        List<PackageableElement> associations = pureModelContextData.getElements().stream().filter(e -> e instanceof org.finos.legend.engine.protocol.pure.v1.model.packageableElement.domain.Association).collect(Collectors.toList());
+        List<PackageableElement> enumerations = pureModelContextData.getElements().stream().filter(e -> e instanceof org.finos.legend.engine.protocol.pure.v1.model.packageableElement.domain.Enumeration).collect(Collectors.toList());
+        List<PackageableElement> functions = pureModelContextData.getElements().stream().filter(e -> e instanceof org.finos.legend.engine.protocol.pure.v1.model.packageableElement.domain.Function).collect(Collectors.toList());
+
+        CompiledExecutionSupport es = pureModel.getExecutionSupport();
+
+        if (classChecks(classes, pureModel, es) && associationChecks(associations, pureModel, es) && enumerationChecks(enumerations, pureModel, es) && functionChecks(functions, pureModel, es))
+        {
+            return Response.ok().build();
+        }
+        else
+        {
+            return Response.status(400).entity("Dataspace does not match the constraints - " + messages).build();
+        }
+    }
+
+    private boolean classChecks(List<PackageableElement> classes, PureModel pureModel, CompiledExecutionSupport es)
```

Comment:
as per comment above, all the code from line 87 to 241 feels like it should belong in Pure code as it is effectively all working on the compiled code so no benefit to having to having this in java

Reply:
This is calling pure functions already and then just aggregating all the responses from the individual checks

Reply:
yes it is calling pure functions, but the entire logic for calling them should also reside in Pure. 
We already have this, so adding a runChecksForDataSpace funtion can be added in similar fashion which then calls the runChecks here.
https://github.com/finos/legend-engine/blob/7677747c788b6bf517e6baea7ab6237d3f220761/legend-engine-xts-analytics/legend-engine-xts-analytics-quality/legend-engine-xt-analytics-quality-pure/src/main/resources/core_analytics_quality/checksEngine.pure#L101


Having code co-located between languages/contexts (resource impl vs funtional impl) vs  is far from ideal
1. The logic has been localized to the implementation of an opinionated rest API whereas the functional implementation is completely generic. e.g. if I wanted to calculate the analytics for a dataspace for another context I cannot do that with this implementation.
2. The tests are mixing concerns/language. json -> rest -> pure -> java. The tests here are really just testing the impl of the check functions themselves and the concat logic. Neither of which make sense in the context of this API really. The check functions already have tests so these tests are not adding anything beneficial. One simple tests for checking instrumentation is fine, but comprehensive logical checking is not required here.

In general
1. try to keep Resource classes to exactly that, to handle the interface of the web call and nothing else. A resource class should be very slim and have very limited/no logic
3. Any instrumentation of pure code from java should be singular in nature. If you are chaining calls together, it is a sign that the implementation likely lives in Pure also.

Reply:
Ah makes sense. Thanks, i will make the changes